File: backend/app/domain/results.py
"""Engine output types.

Everything an engine produces is a plain dataclass here so that services can
persist, serialise and diff results without knowing engine internals. Every
numeric result carries the reasons that produced it (spec Phase 34).
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime
from typing import Any, Literal

from app.domain.enums import (
    AlertSeverity,
    AlertType,
    DataQualityCode,
    DataQualitySeverity,
    ReadinessState,
    RiskLevel,
)

# ---------------------------------------------------------------------------
# Priority
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class FactorScore:
    key: str
    name: str
    kind: Literal["bonus", "penalty"]
    raw_score: float  # 0..100
    weight: float  # normalised 0..1 (0 when disabled)
    points: float  # signed contribution to the final score
    reason: str
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True)
class PriorityAdjustment:
    kind: Literal["aging", "fairness", "expedite", "override", "customer_rule", "erp_priority"]
    points: float
    reason: str
    source_id: str | None = None


@dataclass(slots=True)
class PriorityResult:
    order_id: str
    score: float
    base_score: float
    factors: list[FactorScore]
    adjustments: list[PriorityAdjustment]
    readiness: ReadinessState
    blocked: bool
    blocking_reasons: list[str]
    risk_level: RiskLevel
    explanation: str
    profile_id: str
    profile_version: int
    computed_at: datetime
    hours_until_due: float | None = None
    projected_completion: datetime | None = None
    projected_lateness_hours: float | None = None
    forced_next: bool = False
    rank: int | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


# ---------------------------------------------------------------------------
# Constraints
# ---------------------------------------------------------------------------


@dataclass(slots=True, frozen=True)
class Violation:
    constraint_key: str
    message: str
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True, frozen=True)
class Penalty:
    constraint_key: str
    cost: float  # in "minute-equivalents"; comparable to time
    message: str
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True, frozen=True)
class Blocker:
    state: ReadinessState
    message: str
    resolves_at: datetime | None = None  # when the blocker is expected to clear
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True)
class EligibilityResult:
    operation_id: str
    eligible_machine_ids: list[str]
    rejected: dict[str, list[Violation]]  # machine_id -> violations


# ---------------------------------------------------------------------------
# Scheduling
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class MachineCandidate:
    machine_id: str
    rank: int
    expected_start: datetime | None
    expected_end: datetime | None
    setup_minutes: float
    run_minutes: float
    soft_cost: float
    reasons: list[str]
    recommended: bool = False


@dataclass(slots=True)
class MachineRecommendation:
    order_id: str
    operation_id: str
    eligible: list[MachineCandidate]
    recommended_machine_id: str | None
    rejected: dict[str, list[str]]  # machine_id -> reasons


@dataclass(slots=True)
class ScheduleEntry:
    entry_id: str
    machine_id: str
    order_id: str
    operation_id: str
    sequence_on_machine: int
    setup_start: datetime
    start: datetime
    end: datetime
    setup_minutes: float
    run_minutes: float
    quantity: float
    priority_score: float
    placement_reason: str
    is_last_operation: bool = False
    expected_completion: datetime | None = None  # order-level completion if last op
    due_date: datetime | None = None
    expected_lateness_hours: float | None = None
    locked: bool = False
    batch_key: str | None = None
    setup_family: str | None = None
    material_id: str | None = None
    customer_id: str | None = None


@dataclass(slots=True)
class UnscheduledItem:
    order_id: str
    operation_id: str | None
    reason_code: str
    reason: str
    readiness: ReadinessState | None = None


@dataclass(slots=True)
class ScheduleMetrics:
    scheduled_orders: int = 0
    unscheduled_orders: int = 0
    scheduled_operations: int = 0
    on_time_orders: int = 0
    late_orders: int = 0
    orders_at_risk: int = 0
    on_time_pct: float = 0.0
    avg_lateness_hours: float = 0.0
    max_lateness_hours: float = 0.0
    total_tardiness_hours: float = 0.0
    total_run_hours: float = 0.0
    total_setup_hours: float = 0.0
    setup_count: int = 0
    makespan_hours: float = 0.0
    overall_utilization_pct: float = 0.0
    machine_utilization_pct: dict[str, float] = field(default_factory=dict)
    revenue_scheduled: float = 0.0
    revenue_at_risk: float = 0.0
    margin_at_risk: float = 0.0
    wip_orders_avg: float = 0.0


@dataclass(slots=True)
class ScheduleQuality:
    score: float  # 0..100
    components: dict[str, float]  # component -> 0..100
    weights: dict[str, float]
    summary: str


@dataclass(slots=True)
class ScheduleResult:
    algorithm: str
    algorithm_version: str
    profile_id: str
    profile_version: int
    config_version: int
    generated_at: datetime
    horizon_start: datetime
    horizon_end: datetime
    entries: list[ScheduleEntry] = field(default_factory=list)
    unscheduled: list[UnscheduledItem] = field(default_factory=list)
    metrics: ScheduleMetrics = field(default_factory=ScheduleMetrics)
    quality: ScheduleQuality | None = None
    machine_recommendations: dict[str, MachineRecommendation] = field(default_factory=dict)
    warnings: list[str] = field(default_factory=list)
    run_id: str | None = None

    def entries_for_machine(self, machine_id: str) -> list[ScheduleEntry]:
        return sorted(
            (e for e in self.entries if e.machine_id == machine_id),
            key=lambda e: (e.setup_start, e.sequence_on_machine),
        )

    def entries_for_order(self, order_id: str) -> list[ScheduleEntry]:
        return sorted((e for e in self.entries if e.order_id == order_id), key=lambda e: e.start)

    def order_completion(self) -> dict[str, datetime]:
        out: dict[str, datetime] = {}
        for e in self.entries:
            cur = out.get(e.order_id)
            if cur is None or e.end > cur:
                out[e.order_id] = e.end
        return out


# ---------------------------------------------------------------------------
# Simulation / replanning
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class OrderDelta:
    order_id: str
    baseline_completion: datetime | None
    scenario_completion: datetime | None
    delta_hours: float | None
    baseline_late: bool
    scenario_late: bool
    baseline_machine_id: str | None = None
    scenario_machine_id: str | None = None
    baseline_score: float | None = None
    scenario_score: float | None = None


@dataclass(slots=True)
class ScheduleDiff:
    orders_affected: int
    orders_moved_machine: int
    orders_resequenced: int
    orders_newly_late: int
    orders_newly_on_time: int
    late_orders_before: int
    late_orders_after: int
    on_time_pct_before: float
    on_time_pct_after: float
    avg_lateness_before: float
    avg_lateness_after: float
    utilization_before: float
    utilization_after: float
    setup_hours_before: float
    setup_hours_after: float
    revenue_at_risk_before: float
    revenue_at_risk_after: float
    margin_at_risk_before: float
    margin_at_risk_after: float
    additional_overtime_hours: float
    bottlenecks_before: list[str]
    bottlenecks_after: list[str]
    order_deltas: list[OrderDelta] = field(default_factory=list)
    summary: str = ""


@dataclass(slots=True)
class SimulationResult:
    simulation_id: str
    scenarios: list[dict[str, Any]]
    baseline: ScheduleResult
    scenario: ScheduleResult
    diff: ScheduleDiff
    baseline_priorities: dict[str, float]
    scenario_priorities: dict[str, float]
    generated_at: datetime


@dataclass(slots=True)
class ReplanDecision:
    should_replan: bool
    reason: str
    improvement_pct: float
    changed_entries: int
    frozen_violations: int
    requires_approval: bool
    triggers: list[str] = field(default_factory=list)


# ---------------------------------------------------------------------------
# Analytics
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class CapacityRow:
    dimension: str  # "machine" | "machine_group" | "process" | "department"
    key: str
    period_start: datetime
    period_end: datetime
    required_hours: float
    available_hours: float

    @property
    def gap_hours(self) -> float:
        return self.available_hours - self.required_hours

    @property
    def utilization_pct(self) -> float:
        if self.available_hours <= 0:
            return 100.0 if self.required_hours > 0 else 0.0
        return 100.0 * self.required_hours / self.available_hours


@dataclass(slots=True)
class Bottleneck:
    resource_type: str  # "machine_group" | "machine" | "process" | "material" | "tooling"
    resource_id: str
    resource_name: str
    utilization_pct: float
    orders_waiting: int
    capacity_shortfall_hours: float
    revenue_at_risk: float
    margin_at_risk: float
    severity: RiskLevel
    recommendation: str


@dataclass(slots=True)
class ExecutiveKpis:
    total_open_orders: int
    total_pending_quantity: float
    orders_due_today: int
    orders_due_tomorrow: int
    orders_due_this_week: int
    overdue_orders: int
    at_risk_orders: int
    on_time_delivery_pct: float | None
    expected_on_time_delivery_pct: float | None
    machine_utilization_pct: float | None
    capacity_utilization_pct: float | None
    revenue_at_risk: float
    margin_at_risk: float
    blocked_by_material: int
    blocked_by_tooling: int
    blocked_by_machine: int
    waiting_for_approval: int
    blocked_total: int
    scheduled_orders: int
    unscheduled_orders: int
    as_of: datetime


# ---------------------------------------------------------------------------
# Data quality & alerts
# ---------------------------------------------------------------------------


@dataclass(slots=True)
class DataQualityIssue:
    code: DataQualityCode
    severity: DataQualitySeverity
    entity_type: str
    entity_id: str
    message: str
    field_name: str | None = None
    recommendation: str | None = None
    details: dict[str, Any] = field(default_factory=dict)


@dataclass(slots=True)
class Alert:
    alert_type: AlertType
    severity: AlertSeverity
    title: str
    reason: str
    recommended_action: str
    raised_at: datetime
    order_id: str | None = None
    machine_id: str | None = None
    entity_ref: str | None = None
    dedupe_key: str = ""
    details: dict[str, Any] = field(default_factory=dict)

File: frontend/src/api/types/results.ts
/** Engine outputs (backend/app/domain/results.py) plus schedule versioning and simulation scenarios. */

import type {
  AdjustmentKind,
  AlertSeverity,
  AlertType,
  DataQualityCode,
  DataQualitySeverity,
  FactorKind,
  ReadinessState,
  RiskLevel,
  ScheduleStatus,
  SyncMode,
} from "./enums";

// --------------------------------------------------------------- results

export interface FactorScore {
  key: string;
  name: string;
  kind: FactorKind;
  raw_score: number;
  weight: number;
  points: number;
  reason: string;
  details: Record<string, unknown>;
}

export interface PriorityAdjustment {
  kind: AdjustmentKind;
  points: number;
  reason: string;
  source_id: string | null;
}

export interface PriorityResult {
  order_id: string;
  score: number;
  base_score: number;
  factors: FactorScore[];
  adjustments: PriorityAdjustment[];
  readiness: ReadinessState;
  blocked: boolean;
  blocking_reasons: string[];
  risk_level: RiskLevel;
  explanation: string;
  profile_id: string;
  profile_version: number;
  computed_at: string;
  hours_until_due: number | null;
  projected_completion: string | null;
  projected_lateness_hours: number | null;
  forced_next: boolean;
  rank: number | null;
}

export interface Blocker {
  state: ReadinessState;
  message: string;
  resolves_at: string | null;
  details: Record<string, unknown>;
}

export interface MachineCandidate {
  machine_id: string;
  rank: number;
  expected_start: string | null;
  expected_end: string | null;
  setup_minutes: number;
  run_minutes: number;
  soft_cost: number;
  reasons: string[];
  recommended: boolean;
}

export interface MachineRecommendation {
  order_id: string;
  operation_id: string;
  eligible: MachineCandidate[];
  recommended_machine_id: string | null;
  rejected: Record<string, string[]>;
}

export interface ScheduleEntry {
  entry_id: string;
  machine_id: string;
  order_id: string;
  operation_id: string;
  sequence_on_machine: number;
  setup_start: string;
  start: string;
  end: string;
  setup_minutes: number;
  run_minutes: number;
  quantity: number;
  priority_score: number;
  placement_reason: string;
  is_last_operation: boolean;
  expected_completion: string | null;
  due_date: string | null;
  expected_lateness_hours: number | null;
  locked: boolean;
  batch_key: string | null;
  setup_family: string | null;
  material_id: string | null;
  customer_id: string | null;
}

export interface UnscheduledItem {
  order_id: string;
  operation_id: string | null;
  reason_code: string;
  reason: string;
  readiness: ReadinessState | null;
}

export interface ScheduleMetrics {
  scheduled_orders: number;
  unscheduled_orders: number;
  scheduled_operations: number;
  on_time_orders: number;
  late_orders: number;
  orders_at_risk: number;
  on_time_pct: number;
  avg_lateness_hours: number;
  max_lateness_hours: number;
  total_tardiness_hours: number;
  total_run_hours: number;
  total_setup_hours: number;
  setup_count: number;
  makespan_hours: number;
  overall_utilization_pct: number;
  machine_utilization_pct: Record<string, number>;
  revenue_scheduled: number;
  revenue_at_risk: number;
  margin_at_risk: number;
  wip_orders_avg: number;
}

export interface ScheduleQuality {
  score: number;
  components: Record<string, number>;
  weights: Record<string, number>;
  summary: string;
}

export interface ScheduleResult {
  algorithm: string;
  algorithm_version: string;
  profile_id: string;
  profile_version: number;
  config_version: number;
  generated_at: string;
  horizon_start: string;
  horizon_end: string;
  entries: ScheduleEntry[];
  unscheduled: UnscheduledItem[];
  metrics: ScheduleMetrics;
  quality: ScheduleQuality | null;
  machine_recommendations: Record<string, MachineRecommendation>;
  warnings: string[];
  run_id: string | null;
}

/** schedule_versions row (DESIGN_CONTRACT §10). */
export interface ScheduleVersion {
  version_number: number;
  status: ScheduleStatus;
  generated_at: string;
  generated_by: string | null;
  approved_by?: string | null;
  approved_at?: string | null;
  published_at?: string | null;
  algorithm: string;
  algorithm_version: string;
  profile_id: string;
  profile_version: number;
  config_version: number;
  run_id: string | null;
  input_snapshot_id?: string | null;
  note?: string | null;
  quality_score?: number | null;
}

/** GET /schedule response: a result plus its version envelope (null before any run). */
export interface ScheduleView extends ScheduleResult {
  version: ScheduleVersion | null;
}

export interface OrderDelta {
  order_id: string;
  baseline_completion: string | null;
  scenario_completion: string | null;
  delta_hours: number | null;
  baseline_late: boolean;
  scenario_late: boolean;
  baseline_machine_id: string | null;
  scenario_machine_id: string | null;
  baseline_score: number | null;
  scenario_score: number | null;
}

export interface ScheduleDiff {
  orders_affected: number;
  orders_moved_machine: number;
  orders_resequenced: number;
  orders_newly_late: number;
  orders_newly_on_time: number;
  late_orders_before: number;
  late_orders_after: number;
  on_time_pct_before: number;
  on_time_pct_after: number;
  avg_lateness_before: number;
  avg_lateness_after: number;
  utilization_before: number;
  utilization_after: number;
  setup_hours_before: number;
  setup_hours_after: number;
  revenue_at_risk_before: number;
  revenue_at_risk_after: number;
  margin_at_risk_before: number;
  margin_at_risk_after: number;
  additional_overtime_hours: number;
  bottlenecks_before: string[];
  bottlenecks_after: string[];
  order_deltas: OrderDelta[];
  summary: string;
}

export interface SimulationResult {
  simulation_id: string;
  scenarios: Scenario[];
  baseline: ScheduleResult;
  scenario: ScheduleResult;
  diff: ScheduleDiff;
  baseline_priorities: Record<string, number>;
  scenario_priorities: Record<string, number>;
  generated_at: string;
}

/** Simulation scenarios (DESIGN_CONTRACT §6.5), discriminated on `kind`. */
export type Scenario =
  | { kind: "machine_down"; machine_id: string; start: string; end: string; reason?: string }
  | { kind: "urgent_orders"; order_ids: string[]; boost_points?: number }
  | { kind: "add_machine"; template_machine_id: string; machine_id?: string; machine_name?: string }
  | { kind: "extra_shift"; machine_ids?: string[]; start: string; end: string }
  | { kind: "working_day"; date: string; machine_ids?: string[] }
  | { kind: "outsource"; order_ids: string[] }
  | { kind: "material_delay"; material_id: string; delay_hours: number }
  | { kind: "prioritize_customer"; customer_id: string; boost_points?: number }
  | { kind: "weight_change"; weights: Record<string, number> }
  | { kind: "due_date_change"; order_id: string; new_due_date: string };

export type ScenarioKind = Scenario["kind"];

export interface CapacityRow {
  dimension: "machine" | "machine_group" | "process" | "department";
  key: string;
  period_start: string;
  period_end: string;
  required_hours: number;
  available_hours: number;
  gap_hours?: number;
  utilization_pct?: number;
}

export interface Bottleneck {
  resource_type: "machine_group" | "machine" | "process" | "material" | "tooling";
  resource_id: string;
  resource_name: string;
  utilization_pct: number;
  orders_waiting: number;
  capacity_shortfall_hours: number;
  revenue_at_risk: number;
  margin_at_risk: number;
  severity: RiskLevel;
  recommendation: string;
}

export interface ExecutiveKpis {
  total_open_orders: number;
  total_pending_quantity: number;
  orders_due_today: number;
  orders_due_tomorrow: number;
  orders_due_this_week: number;
  overdue_orders: number;
  at_risk_orders: number;
  on_time_delivery_pct: number | null;
  expected_on_time_delivery_pct: number | null;
  machine_utilization_pct: number | null;
  capacity_utilization_pct: number | null;
  revenue_at_risk: number;
  margin_at_risk: number;
  blocked_by_material: number;
  blocked_by_tooling: number;
  blocked_by_machine: number;
  waiting_for_approval: number;
  blocked_total: number;
  scheduled_orders: number;
  unscheduled_orders: number;
  as_of: string;
}

/** One period of on-time-delivery history (GET /analytics/on-time-delivery). */
export interface OtdPoint {
  period_start: string;
  period_end: string;
  orders_due: number;
  on_time: number;
  late: number;
  on_time_pct: number | null;
}

/** DataQualityIssueResponse: one issue of the latest data-quality run (GET /data-quality/issues). */
export interface DataQualityIssue {
  issue_id: string;
  run_id: string;
  detected_at: string;
  /** Serialised enum value (DataQualityCode). */
  code: DataQualityCode;
  /** Serialised enum value (DataQualitySeverity). */
  severity: DataQualitySeverity;
  entity_type: string;
  entity_id: string;
  message: string;
  field_name: string | null;
  recommendation: string | null;
  details: Record<string, unknown>;
}

/** AlertResponse (GET /alerts). */
export interface Alert {
  alert_id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  title: string;
  reason: string;
  recommended_action: string;
  raised_at: string;
  last_seen_at: string;
  occurrences: number;
  active: boolean;
  order_id: string | null;
  machine_id: string | null;
  entity_ref: string | null;
  details: Record<string, unknown>;
  acknowledged: boolean;
  acknowledged_by: string | null;
  acknowledged_at: string | null;
  resolved_at: string | null;
}

/** AuditEntryResponse: one audit_log row (GET /audit, OrderDetail.audit). */
export interface AuditEntry {
  audit_id: string;
  user_id: string;
  timestamp: string;
  entity_type: string;
  entity_id: string;
  action: string;
  previous_value: Record<string, unknown> | unknown[] | null;
  new_value: Record<string, unknown> | unknown[] | null;
  reason: string | null;
  request_id: string | null;
  details: Record<string, unknown>;
}

export interface SyncRun {
  run_id: string;
  mode: SyncMode;
  started_at: string;
  finished_at: string | null;
  status: string;
  records_fetched?: number;
  records_upserted?: number;
  errors?: string[];
  message?: string | null;
}

export interface HealthResponse {
  status: string;
  database: string;
  version: string;
  environment: string;
  time: string;
}
